File: DeepspaceRuby/lib/NumericDamage.rb
#encoding: UTF-8

require_relative 'Damage'
require_relative 'NumericDamageToUI'

module Deepspace
    class NumericDamage < Damage

        def initialize(nW,sB)
            super(sB)
            @nWeapons = nW
        end

        def copy
            return NumericDamage.new(@nWeapons,nShields)
        end

        def getUIversion
            return NumericDamageToUI.new(self)
        end

        def adjust(w,s)
            minNShields = [@nShields,s.length].min
            minNWeapons = [@nWeapons,w.length].min
            
            return NumericDamage.new(minNWeapons,minNShields)
        end

        def nWeapons
            return @nWeapons
        end

        def discardWeapon(w)
            if(@nWeapons > 0)
                @nWeapons-=1	
            end
        end

        def hasNoEffect
            return @nWeapons == 0 && super
        end

        def to_s
            out = super
            out += "nWeapons: #{@nWeapons} "
        end
    end   
end

File: deepspace/NumericDamage.java
/*
 * @author: Jesús García León
 * @file: 
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author usuario
 */
public class NumericDamage extends Damage{
    private int nWeapons;
    
    NumericDamage(int w, int s){
        super(s);
        nWeapons = w;
    }

    @Override
    public NumericDamageToUI getUIversion() {
        return new NumericDamageToUI(this);
    }

    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int minNShield = Math.min(nShields, s.size());
        int minNWeapons = Math.min(nWeapons, w.size());
        
        return new NumericDamage(minNWeapons, minNShield);
    }

    @Override
    public void discardWeapon(Weapon w) {
        if (nWeapons > 0)
            nWeapons--;
    }

    @Override
    public boolean hasNoEffect() {
        return ((nWeapons == 0) && (nShields == 0));
    }

    public int getNWeapons() {
        return nWeapons;
    }

    Damage copy(NumericDamage d) {
        super.copy(d);
        nWeapons= d.nWeapons;
        return this;
    }
    
    @Override
    public String toString(){
        String aux = super.toString();
        aux += "\nnWeapons: " + nWeapons + "\n";
        return aux;
    }
    
}
